dlt/destinations/impl/filesystem/filesystem.py

```diff
@@ -220,3 +239,126 @@ def __exit__(
+        # TODO: this is also duplicate across all destinations. DRY this.
+        load_data = {
+            "load_id": load_id,
+            "schema_name": self.schema.name,
+            "status": 0,
+            "inserted_at": pendulum.now().isoformat(),
+            "schema_version_hash": self.schema.version_hash,
+        }
+        filepath = (
+            f"{self.dataset_path}/{self.schema.loads_table_name}/{self.schema.name}.{load_id}.jsonl"
+        )
+
+        self._write_to_json_file(filepath, load_data)
+
+    #
+    # state read/write
+    #
+
+    def _get_state_file_name(self, pipeline_name: str, version_hash: str) -> str:
+        """gets full path for schema file for a given hash"""
+        safe_hash = "".join(
+            [c for c in version_hash if re.match(r"\w", c)]
+        )  # remove all special chars from hash
+        return (
+            f"{self.dataset_path}/{self.schema.state_table_name}/{pipeline_name}__{safe_hash}.jsonl"
+        )
+
+    def store_current_state(self) -> None:
+        # get state doc from current pipeline
+        from dlt.common.configuration.container import Container
+        from dlt.common.pipeline import PipelineContext
+        from dlt.pipeline.state_sync import state_doc
+
+        pipeline = Container()[PipelineContext].pipeline()
+        state = pipeline.state
+        doc = state_doc(state)
+
+        # get paths
+        current_path = self._get_state_file_name(pipeline.pipeline_name, "current")
+        hash_path = self._get_state_file_name(
+            pipeline.pipeline_name, self.schema.stored_version_hash
+        )
+
+        # write
+        self._write_to_json_file(current_path, doc)
+        self._write_to_json_file(hash_path, doc)
+
+    def get_stored_state(self, pipeline_name: str) -> Optional[StateInfo]:
+        # raise if dir not initialized
+        filepath = self._get_state_file_name(pipeline_name, "current")
+        dirname = os.path.dirname(filepath)
+        if not self.fs_client.isdir(dirname):
+            raise DestinationUndefinedEntity({"dir": dirname})
+
+        """Loads compressed state from destination storage"""
+        if self.fs_client.exists(filepath):
+            state_json = json.loads(self.fs_client.read_text(filepath))
+            state_json.pop("version_hash")
+            return StateInfo(**state_json)
+
+        return None
+
+    #
+    # Schema read/write
+    #
+
+    def _get_schema_file_name(self, version_hash: str) -> str:
+        """gets full path for schema file for a given hash"""
+        safe_hash = "".join(
+            [c for c in version_hash if re.match(r"\w", c)]
+        )  # remove all special chars from hash
+        return f"{self.dataset_path}/{self.schema.version_table_name}/{self.schema.name}__{safe_hash}.jsonl"
+
+    def get_stored_schema(self) -> Optional[StorageSchemaInfo]:
+        """Retrieves newest schema from destination storage"""
+        return self.get_stored_schema_by_hash("current")
```

Comment:
same thing like in state: find the oldest load id

Reply:
done (assuming you mean the newest load id :) )